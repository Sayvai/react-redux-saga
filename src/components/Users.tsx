import React, { useEffect } from "react";
import { connect } from "react-redux";
import { getUsersStart as getUsersStartAction } from "../store/users/users.actions";
import { selectGetUsersData } from "../store/users/users.selectors";
import { User } from "../store/users/users.types";
import styled from "styled-components";

interface IUsers {
  users: User[];
  limit?: number;
  getUsersStart: () => void;
}

const Ul = styled.ul`
  display: flex;
  justify-content: space-around;
  gap: 2em;
  list-style-type: none;
`;

Ul.displayName = "UL";

const Li = styled.li`
  background: #ffe5e5;
  color: darkred;
`;

Li.displayName = "LI";

const Users: React.FC<IUsers> = ({ users, limit = 0, getUsersStart }) => {
  useEffect(() => {
    getUsersStart();
  }, [getUsersStart]);

  const renderUsers = () => {
    const renderLimit = limit === 0 ? users.length : limit;
    const renderableUserList = users.slice(0, renderLimit);

    return renderableUserList.map((user) => (
      <Li aria-label={user.name} data-testid="user" key={user.id}>
        {user.name}
      </Li>
    ));
  };

  return (
    <div data-testid="users-wrapper">
      <h1>List of users (max limit: {limit}):</h1>
      {users.length > 0 && (
        <Ul aria-label="users list" data-testid="users-list">
          {renderUsers()}
        </Ul>
      )}
    </div>
  );
};

const mapStateToProps = (state) => ({
  users: selectGetUsersData(state),
});

const actionCreators = {
  getUsersStart: getUsersStartAction,
};

export default React.memo(connect(mapStateToProps, actionCreators)(Users));
