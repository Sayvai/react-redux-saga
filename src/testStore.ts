import { createStore, combineReducers } from "redux";
import users from "./store/users/users.reducer";

export const createTestStore = (mockStoreData = {}) => {
  const store = createStore(
    combineReducers({
      users: () => ({ userList: [{ id: 1, name: "Lionel Messi" }] }),
      ...mockStoreData,
    })
  );
  return store;
};
