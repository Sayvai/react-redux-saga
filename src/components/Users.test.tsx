import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { createTestStore } from "../testStore";
import Users from "./Users";

describe("<Users/>", () => {
  let store;
  let props;

  beforeEach(() => {
    const storeUsers = () => ({
      userList: [
        { id: 1, name: "Lionel Messi" },
        { id: 2, name: "Cristiano Ronaldo" },
        { id: 3, name: "Thierry Henry" },
        { id: 4, name: "Martin Ã˜degaard" },
        { id: 5, name: "Tony Adams" },
        { id: 6, name: "Cesc Fabregas" },
        { id: 7, name: "Kylian Mbappe" },
        { id: 8, name: "Harry Kane" },
        { id: 9, name: "Kevin De Bruyne" },
        { id: 10, name: "Mohammed Salah" },
        { id: 11, name: "Kai Havertz" },
      ],
    });

    store = createTestStore({ users: storeUsers });

    props = {
      limit: 10,
      getUsersStart: jest.fn(),
    };

    jest.spyOn(React, "useEffect").mockImplementation((cb) => cb());
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should render the full list of available users", async () => {
    // given
    render(
      <Provider store={store}>
        <Users {...props} />
      </Provider>
    );

    // then
    const usersWrapper = screen.getByTestId("users-wrapper");
    const usersList = screen.getByTestId("users-list");
    const user = await screen.findAllByTestId("user");

    expect(usersWrapper).toMatchSnapshot();
    expect(usersList).toBeInTheDocument();
    expect(user).toHaveLength(10);
  });

  it("should render only 5 users from the list", async () => {
    // given
    render(
      <Provider store={store}>
        <Users {...props} limit={5} />
      </Provider>
    );

    // then
    const usersWrapper = screen.getByTestId("users-wrapper");
    const usersList = screen.getByTestId("users-list");
    const user = await screen.findAllByTestId("user");

    expect(usersWrapper).toMatchSnapshot();
    expect(usersList).toBeInTheDocument();
    expect(user).toHaveLength(5);
  });

  // TODO - Mocked dispatch function not reflected as a mock jest function during component test run. May need to think about mocking the store using redux-mock-store ðŸ¤”
  it.skip("should have called store dispatch function props.getUsersStart() on intiialisation", () => {
    // given
    render(
      <Provider store={store}>
        <Users {...props} />
      </Provider>
    );

    // then
    console.log("TEST props.getUsersStart()", props.getUsersStart);
    expect(props.getUsersStart).toHaveBeenCalled();
  });
});
