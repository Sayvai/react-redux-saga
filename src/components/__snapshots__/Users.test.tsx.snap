// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users/> should render only 5 users from the list 1`] = `
<div
  data-testid="users-wrapper"
>
  <h1>
    List of users (max limit: 
    5
    ):
  </h1>
  <ul
    aria-label="users list"
    class="sc-bcXHqe ctVCar"
    data-testid="users-list"
  >
    <li
      aria-label="Lionel Messi"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Lionel Messi
    </li>
    <li
      aria-label="Cristiano Ronaldo"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Cristiano Ronaldo
    </li>
    <li
      aria-label="Thierry Henry"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Thierry Henry
    </li>
    <li
      aria-label="Martin Ødegaard"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Martin Ødegaard
    </li>
    <li
      aria-label="Tony Adams"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Tony Adams
    </li>
  </ul>
</div>
`;

exports[`<Users/> should render the full list of available users 1`] = `
<div
  data-testid="users-wrapper"
>
  <h1>
    List of users (max limit: 
    10
    ):
  </h1>
  <ul
    aria-label="users list"
    class="sc-bcXHqe ctVCar"
    data-testid="users-list"
  >
    <li
      aria-label="Lionel Messi"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Lionel Messi
    </li>
    <li
      aria-label="Cristiano Ronaldo"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Cristiano Ronaldo
    </li>
    <li
      aria-label="Thierry Henry"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Thierry Henry
    </li>
    <li
      aria-label="Martin Ødegaard"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Martin Ødegaard
    </li>
    <li
      aria-label="Tony Adams"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Tony Adams
    </li>
    <li
      aria-label="Cesc Fabregas"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Cesc Fabregas
    </li>
    <li
      aria-label="Kylian Mbappe"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Kylian Mbappe
    </li>
    <li
      aria-label="Harry Kane"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Harry Kane
    </li>
    <li
      aria-label="Kevin De Bruyne"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Kevin De Bruyne
    </li>
    <li
      aria-label="Mohammed Salah"
      class="sc-gswNZR fkIaOi"
      data-testid="user"
    >
      Mohammed Salah
    </li>
  </ul>
</div>
`;
